### ISSUE TITLE
Incorrect credentials are cached?

### ISSUE THREAD

**Issue by yanokwa** [opendatakit/briefcase#334]
This is an issue I see with v1.9.beta-0.

To reproduce, pull from a server using bad credentials. Now, use the correct credentials. It'll still fail. Now quit, use the correct credentials, and it'll work. 

I believe the reverse is true as well. If I pull using correct credentials, then change out the connections and try to pull using the wrong credentials, it works when it should fail.

**Comment by ggalmazor** [comment]
I'm trying to trace back this problem. It appears to be from before the changes for the new Export tab...

**Comment by ggalmazor** [comment]
I can reproduce this as far as commit #c5c9bc19: Race condition using parallel pulls causes HTTP 401 Unauthorized (#206)

This is the diff between that commit and the previous one:
```diff
diff --git a/src/org/opendatakit/briefcase/model/ServerConnectionInfo.java b/src/org/opendatakit/briefcase/model/ServerConnectionInfo.java
index dade91c..c4833fb 100644
--- a/src/org/opendatakit/briefcase/model/ServerConnectionInfo.java
+++ b/src/org/opendatakit/briefcase/model/ServerConnectionInfo.java
@@ -21,12 +21,13 @@ import org.apache.http.client.protocol.HttpClientContext;
 import org.opendatakit.briefcase.util.WebUtils;
 
 public class ServerConnectionInfo {
+
   private String url;
   private final String token; // for legacy ODK Aggregate 0.9.8 access
   private final String username;
   private final char[] password;
   private final boolean isOpenRosaServer;
-  private HttpClientContext httpContext = null;
+  private static final ThreadLocal<HttpClientContext> threadSafeContext = new ThreadLocal<>();
 
   public ServerConnectionInfo(String url, String username, char[] cs) {
     this.url = url;
@@ -69,11 +70,11 @@ public class ServerConnectionInfo {
   }
 
   public HttpClientContext getHttpContext() {
-    return httpContext;
+    return threadSafeContext.get();
   }
 
   public void setHttpContext(HttpClientContext httpContext) {
-    this.httpContext = httpContext;
+    this.threadSafeContext.set(httpContext);
   }
 
   public boolean isOpenRosaServer() {
``` 

@batkinson, you were working on Briefcase those dates... I know much time has passed since then but maybe you know some easy fix for this one.

**Comment by ggalmazor** [comment]
@opendatakit-bot unclaim

**Comment by ggalmazor** [comment]
This one is tricky to solve in a contained manner. We're making `HttpClientContext` thread-safe, meaning that a thread will reuse the same instance, which has lead to the accidental coupling of different business operations: testing the server connection details, getting the remote form list and pulling submissions.

What we want is to avoid coupling problems between these operations, i.e., failing a connection test should not affect the next test, but there are more possible scenarios of such coupling problems.

We would achieve a solution to this problem by redesigning these operations to be self-contained. That would require some extra work to break the use of static APIs that currently are facilitating the coupling problem I've described.

I've already prepared a simple fix for this issue, but I'd like for your feedback (@yanokwa, @lognaturel, @dcbriccetti) on maybe investing some more time to improve the design and try to solve the coupling problem at least on the operations from the pull tab. Doing so would lay the ground for further refactors (hopefully "contributor friendly" after this) to solve other similar problems.
